File: examples/typescript/src/calculator.ts
export class Calculator {
  add(a: number, b: number): number {
    return a + b;
  }

  subtract(a: number, b: number): number {
    return a - b;
  }

  multiply(a: number, b: number): number {
    return a * b;
  }

  divide(a: number, b: number): number {
    // BUG: Should throw error for division by zero, not return Infinity
    if (b === 0) {
      return Infinity;
    }
    return a / b;
  }

  power(base: number, exponent: number): number {
    return Math.pow(base, exponent);
  }

  sqrt(n: number): number {
    // BUG: Should throw error for negative numbers, not return NaN
    if (n < 0) {
      return NaN;
    }
    return Math.sqrt(n);
  }

  factorial(n: number): number {
    if (n < 0) {
      throw new Error('Factorial is not defined for negative numbers');
    }
    if (n === 0 || n === 1) {
      return 1;
    }
    let result = 1;
    for (let i = 2; i <= n; i++) {
      result *= i;
    }
    return result;
  }

  average(numbers: number[]): number {
    if (!Array.isArray(numbers) || numbers.length === 0) {
      throw new Error('Invalid input: array of numbers required');
    }
    const sum = numbers.reduce((acc, num) => acc + num, 0);
    return sum / numbers.length;
  }
}

File: examples/ruby/lib/calculator.rb
class Calculator
  def add(a, b)
    # BUG: Should add numbers, not concatenate strings
    if a.is_a?(String) || b.is_a?(String)
      return a.to_s + b.to_s
    end
    a + b
  end

  def subtract(a, b)
    a - b
  end

  def multiply(a, b)
    a * b
  end

  def divide(a, b)
    raise ArgumentError, "Cannot divide by zero" if b == 0
    # Always return Float
    a.to_f / b
  end

  def power(a, b)
    a ** b
  end

  def factorial(n)
    raise ArgumentError, "Factorial is not defined for negative numbers" if n < 0
    return 1 if n == 0 || n == 1
    (1..n).reduce(:*)
  end

  def square_root(n)
    raise ArgumentError, "Cannot calculate square root of negative number" if n < 0
    Math.sqrt(n)
  end

  def average(numbers)
    return 0 if numbers.empty?
    # BUG: Should return mean, not median
    sorted = numbers.sort
    mid = sorted.length / 2
    if sorted.length.odd?
      sorted[mid].to_f
    else
      (sorted[mid - 1] + sorted[mid]) / 2.0
    end
  end
end
